cmd/zc_enumerator.go

```diff
@@ -202,11 +242,62 @@ func initResourceTraverser(source string, location common.Location, ctx *context
 			return nil, errors.New("a valid credential and context must be supplied to create a blobFS traverser")
 		}
 
-		output = newBlobFSTraverser(sourceURL, *p, *ctx, recursive, incrementEnumerationCounter)
+		bfsURL := azbfs.NewBfsURLParts(*sourceURL)
+
+		if bfsURL.FileSystemName == "" || strings.Contains(bfsURL.FileSystemName, "*") {
+			// TODO service traverser
+
+			if !recursive {
+				return nil, errors.New(accountTraversalInherentlyRecursiveError)
+			}
+
+			output = newBlobFSAccountTraverser(sourceURL, *p, *ctx, incrementEnumerationCounter)
+		} else {
+			output = newBlobFSTraverser(sourceURL, *p, *ctx, recursive, incrementEnumerationCounter)
+		}
+	case common.ELocation.S3():
+		sourceURL, err := url.Parse(source)
+		if err != nil {
+			return nil, err
+		}
+
+		s3URLParts, err := common.NewS3URLParts(*sourceURL)
+		if err != nil {
+			return nil, err
+		}
+
+		if ctx == nil {
+			return nil, errors.New("a valid context must be supplied to create a S3 traverser")
+		}
+
+		if s3URLParts.BucketName == "" || strings.Contains(s3URLParts.BucketName, "*") {
+			// TODO convert to path style URL
```

Comment:
Ditto

Reply:
AWS allows two types of URLs depending on how recently the region was made available.

You have the classic hostname style:
bucket.s3.amazonaws.com

And you have the path style:
region.s3.amazonaws.com/bucket

The tricky bit is: The hostname style doesn't require a region, but the path style does.

Reply:
Regions launched since a certain date (I forget exactly when, I think it was sometime in 2018?) can ONLY use the path style.